src/utils/parse.rs

```diff
@@ -0,0 +1,186 @@
+            .or_else(lookup_by_nickname)
+            .cloned()
+            .ok_or(MemberParseError::NotFoundOrMalformed)
+    }
+}
+
+/// Error that can be returned from [`Message::parse`].
+#[derive(Debug)]
+pub enum MessageParseError {
+    Malformed,
+    Http(SerenityError),
+    /// When the `gateway` feature is disabled and the required information was not in cache
+    HttpNotAvailable,
+}
+
+impl std::error::Error for MessageParseError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match self {
+            Self::Malformed => None,
+            Self::Http(e) => Some(e),
+            Self::HttpNotAvailable => None,
+        }
+    }
+}
+
+impl std::fmt::Display for MessageParseError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Malformed => {
+                write!(f, "provided string did not adhere to any known guild message format")
+            },
+            Self::Http(e) => write!(f, "failed to request message data via HTTP: {}", e),
+            Self::HttpNotAvailable => write!(
+                f,
+                "gateway feature is disabled and the required information was not in cache"
+            ),
+        }
+    }
+}
+
+/// Look up a message by a string.
+///
+/// The lookup strategy is as follows (in order):
+/// 1. Lookup by "{channel ID}-{message ID}" (retrieved by shift-clicking on "Copy ID")
+/// 2. Lookup by message ID (the message must be in the context channel)
+/// 3. Lookup by message URL
+#[async_trait::async_trait]
+impl Parse for Message {
+    type Err = MessageParseError;
+
+    #[doc = ""] // to get impl documentation to show
+    async fn parse(ctx: &Context, msg: &Message, s: &str) -> Result<Self, Self::Err> {
+        let extract_from_id_pair = || {
+            let mut parts = s.splitn(2, '-');
+            let channel_id = ChannelId(parts.next()?.parse().ok()?);
+            let message_id = MessageId(parts.next()?.parse().ok()?);
+            Some((channel_id, message_id))
+        };
+
+        let extract_from_message_id = || Some((msg.channel_id, MessageId(s.parse().ok()?)));
+
+        let extract_from_message_url = || {
+            let mut parts = s.strip_prefix("https://discord.com/channels/")?.splitn(3, '/');
+            let _guild_id = GuildId(parts.next()?.parse().ok()?);
+            let channel_id = ChannelId(parts.next()?.parse().ok()?);
+            let message_id = MessageId(parts.next()?.parse().ok()?);
+            Some((channel_id, message_id))
+        };
+
+        let (channel_id, message_id) = extract_from_id_pair()
+            .or_else(extract_from_message_id)
+            .or_else(extract_from_message_url)
+            .ok_or(MessageParseError::Malformed)?;
+
+        #[cfg(feature = "cache")]
+        if let Some(msg) = ctx.cache.message(channel_id, message_id).await {
+            return Ok(msg);
+        }
+
+        if cfg!(feature = "gateway") {
```

Comment:
The `gateway` feature implies the `http` feature, but I don't see why I we couldn't depend on `http` directly. 
```suggestion
        if cfg!(feature = "http") {
```

Reply:
That's just my unfamiliarity with serenity's features (I just use default features for my own bots :sweat_smile:) Thanks for the correction